### DISCUSSION TITLE
Can Generics help with ent code generate?

### DISCUSSION THREAD

**Discussion by FingerLiu** [ent/ent#2234] [Ideas]
I use ent created a project with 10 models.
And I found the binary file generated is over 40Mb.
Can golang generics help to make ent generate less code?
If this is possbile, I would like to have a try.

**Comment by a8m** [comment]
Hey @FingerLiu 👋
I actually answered this question in our [Slack channel](https://gophers.slack.com/archives/C01FMSQDT53/p1640098742444400?thread_ts=1640098204.444100&cid=C01FMSQDT53) a few days ago, so you're welcome to join (see [instructions](https://entgo.io/docs/slack/)) and see my response [there](https://gophers.slack.com/archives/C01FMSQDT53/p1640098742444400?thread_ts=1640098204.444100&cid=C01FMSQDT53). 

But in short, I worked on an experiment with ent with generics, and generics improves so many parts and reduces the amount of generated code. I plan to publish some of the work (to a public branch) after v0.10.0 release (coming soon). But of course, I'll keep it experimental until we'll have enough confidence and feedback from the community.
However, some extensions/plugin can already use it, like:

```go
for _, n := range ent.Iterate(<Any Query>, opts...) {
    // ...
}

ent.Union(
    a.QueryFriends(),
    b.QueryFriends(),
).
QueryPosts().
```

Now, regarding your use case of 10 models generating 40MB file - I'm pretty sure it's unrelated to ent. I can give you some numbers, and you can ping me on Slack if you need some help with the research.
1. A simple hello world app on my laptop with standard build (`go build`) - 1.8M
2. Same app, but with MySQL driver (and `database/sql`) package - 4.9M
3. Added to 2, n base ent project with one entity (with 5 fields) - 5.2M
4. Added additional 10 entities with a few fields for each one (the main file interacts with all of them) - 5.3M
5. Import the schemas in `entgo.io/ent/entc/integration/ent` package with all its dependencies (almost all std packages) - 15MB

There are some tools that can help analyze it, so I'll be happy to help and improve ent if you find anything.

**Reply by FingerLiu** [reply]
Cool! It is really great job, is there something like pre release branch of this? I'd like to have a try.,

**Comment by FingerLiu** [comment]
Did some research and test, it turns out that swagger docs took the most place.
After turn swagger off, my package size get down to 20M and ent was about 2M


![image](https://user-images.githubusercontent.com/3981064/147386392-6c4582db-73c4-4c6e-8d34-32e99fb45347.png)

**Reply by hi019** [reply]
Hey. What tool did you use to generate this?

**Reply by FingerLiu** [reply]
@hi019 https://github.com/knz/go-binsize-viz
